File: Assets/WebGLTemplates/Main/WebRTC/AWebRTCPeer.js
var AWebRtcPeer = function() {
    function e(e) {
        this.mState = WebRtcPeerState.Invalid;
        this.mRtcInternalState = WebRtcInternalState.None;
        this.mIncomingSignalingQueue = new Queue;
        this.mOutgoingSignalingQueue = new Queue;
        this.mDidSendRandomNumber = false;
        this.mRandomNumerSent = 0;
        this.mOfferOptions = {
            offerToReceiveAudio: 0,
            offerToReceiveVideo: 0
        };
        this.gConnectionConfig = {
            optional: [{
                DtlsSrtpKeyAgreement: true
            }]
        };
        this.SetupPeer(e);
        this.OnSetup();
        this.mState = WebRtcPeerState.Created
    }
    e.prototype.GetState = function() {
        return this.mState
    };
    e.prototype.SetupPeer = function(e) {
        var t = this;
        var n = {
            iceServers: [{
                urls: "stun:stun.services.mozilla.com"
            }]
        };
        n.iceServers = new Array;
        for (var i = 0; i < e.length; i++) {
            var o = {
                urls: e[0]
            };
            n.iceServers.push(o)
        }
        var r = window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection;
        this.mPeer = new r(n, this.gConnectionConfig);
        this.mPeer.onicecandidate = function(e) {
            t.OnIceCandidate(e)
        };
        this.mPeer.oniceconnectionstatechange = function(e) {
            t.OnIceConnectionChange()
        };
        this.mPeer.onnegotiationneeded = function(e) {
            t.OnRenegotiationNeeded()
        };
        this.mPeer.onsignalingstatechange = function(e) {
            t.OnSignalingChange()
        }
    };
    e.prototype.DisposeInternal = function() {
        this.Cleanup()
    };
    e.prototype.Dispose = function() {
        if (this.mPeer != null) {
            this.DisposeInternal()
        }
    };
    e.prototype.Cleanup = function() {
        if (this.mState == WebRtcPeerState.Closed || this.mState == WebRtcPeerState.Closing) {
            return
        }
        this.mState = WebRtcPeerState.Closing;
        this.OnCleanup();
        if (this.mPeer != null) this.mPeer.close();
        this.mState = WebRtcPeerState.Closed
    };
    e.prototype.Update = function() {
        if (this.mState != WebRtcPeerState.Closed && this.mState != WebRtcPeerState.Closing && this.mState != WebRtcPeerState.SignalingFailed) this.UpdateState();
        if (this.mState == WebRtcPeerState.Signaling || this.mState == WebRtcPeerState.Created) this.HandleIncomingSignaling()
    };
    e.prototype.UpdateState = function() {
        if (this.mRtcInternalState == WebRtcInternalState.Closed) {
            this.Cleanup()
        } else if (this.mRtcInternalState == WebRtcInternalState.SignalingFailed) {
            this.mState = WebRtcPeerState.SignalingFailed
        } else if (this.mRtcInternalState == WebRtcInternalState.Connected) {
            this.mState = WebRtcPeerState.Connected
        }
    };
    e.prototype.HandleIncomingSignaling = function() {
        while (this.mIncomingSignalingQueue.Count() > 0) {
            var e = this.mIncomingSignalingQueue.Dequeue();
            var t = Helper.tryParseInt(e);
            if (t != null) {
                if (this.mDidSendRandomNumber) {
                    if (t < this.mRandomNumerSent) {
                        SLog.L("Signaling negotiation complete. Starting signaling.");
                        this.StartSignaling()
                    } else if (t == this.mRandomNumerSent) {
                        this.NegotiateSignaling()
                    } else {
                        SLog.L("Signaling negotiation complete. Waiting for signaling.")
                    }
                } else {}
            } else {
                var n = JSON.parse(e);
                if (n.sdp) {
                    var i = new RTCSessionDescription(n);
                    if (i.type == "offer") {
                        this.CreateAnswer(i)
                    } else {
                        this.RecAnswer(i)
                    }
                } else {
                    var o = new RTCIceCandidate(n);
                    if (o != null) {
                        this.mPeer.addIceCandidate(o, function() {}, function(e) {
                            Debug.LogError(e)
                        })
                    }
                }
            }
        }
    };
    e.prototype.AddSignalingMessage = function(e) {
        this.mIncomingSignalingQueue.Enqueue(e)
    };
    e.prototype.DequeueSignalingMessage = function(e) {
        {
            if (this.mOutgoingSignalingQueue.Count() > 0) {
                e.val = this.mOutgoingSignalingQueue.Dequeue();
                return true
            } else {
                e.val = null;
                return false
            }
        }
    };
    e.prototype.EnqueueOutgoing = function(e) {
        {
            this.mOutgoingSignalingQueue.Enqueue(e)
        }
    };
    e.prototype.StartSignaling = function() {
        this.OnStartSignaling();
        this.CreateOffer()
    };
    e.prototype.NegotiateSignaling = function() {
        var e = Random.getRandomInt(0, 2147483647);
        this.mRandomNumerSent = e;
        this.mDidSendRandomNumber = true;
        this.EnqueueOutgoing("" + e)
    };
    e.prototype.CreateOffer = function() {
        var e = this;
        Debug.Log("CreateOffer");
        this.mPeer.createOffer(function(t) {
            var n = JSON.stringify(t);
            e.mPeer.setLocalDescription(t, function() {
                e.RtcSetSignalingStarted();
                e.EnqueueOutgoing(n)
            }, function(t) {
                Debug.LogError(t);
                e.RtcSetSignalingFailed()
            })
        }, function(t) {
            Debug.LogError(t);
            e.RtcSetSignalingFailed()
        }, this.mOfferOptions)
    };
    e.prototype.CreateAnswer = function(e) {
        var t = this;
        Debug.Log("CreateAnswer");
        this.mPeer.setRemoteDescription(e, function() {
            t.mPeer.createAnswer(function(e) {
                var n = JSON.stringify(e);
                t.mPeer.setLocalDescription(e, function() {
                    t.RtcSetSignalingStarted();
                    t.EnqueueOutgoing(n)
                }, function(e) {
                    Debug.LogError(e);
                    t.RtcSetSignalingFailed()
                })
            }, function(e) {
                Debug.LogError(e);
                t.RtcSetSignalingFailed()
            })
        }, function(e) {
            Debug.LogError(e);
            t.RtcSetSignalingFailed()
        })
    };
    e.prototype.RecAnswer = function(e) {
        var t = this;
        Debug.Log("RecAnswer");
        this.mPeer.setRemoteDescription(e, function() {}, function(e) {
            Debug.LogError(e);
            t.RtcSetSignalingFailed()
        })
    };
    e.prototype.RtcSetSignalingStarted = function() {
        if (this.mRtcInternalState == WebRtcInternalState.None) {
            this.mRtcInternalState = WebRtcInternalState.Signaling
        }
    };
    e.prototype.RtcSetSignalingFailed = function() {
        this.mRtcInternalState = WebRtcInternalState.SignalingFailed
    };
    e.prototype.RtcSetConnected = function() {
        if (this.mRtcInternalState == WebRtcInternalState.Signaling) this.mRtcInternalState = WebRtcInternalState.Connected
    };
    e.prototype.RtcSetClosed = function() {
        if (this.mRtcInternalState == WebRtcInternalState.Connected) this.mRtcInternalState = WebRtcInternalState.Closed
    };
    e.prototype.OnIceCandidate = function(e) {
        if (e && e.candidate) {
            var t = e.candidate;
            var n = JSON.stringify(t);
            this.EnqueueOutgoing(n)
        }
    };
    e.prototype.OnIceConnectionChange = function() {
        Debug.Log(this.mPeer.iceConnectionState);
        if (this.mPeer.iceConnectionState == "failed") {
            this.mState = WebRtcPeerState.SignalingFailed
        }
    };
    e.prototype.OnIceGatheringChange = function() {
        Debug.Log(this.mPeer.iceGatheringState)
    };
    e.prototype.OnRenegotiationNeeded = function() {};
    e.prototype.OnSignalingChange = function() {
        Debug.Log(this.mPeer.signalingState);
        if (this.mPeer.signalingState == "closed") {
            this.RtcSetClosed()
        }
    };
    return e
}();

File: Assets/WebRtcNetwork/Byn.Net.Native/AWebRtcPeer.cs
using Byn.Common;
using System;
using System.Collections.Generic;
using UnityEngine;
using WebRtcCSharp;

namespace Byn.Net.Native
{
    public abstract class AWebRtcPeer : IDisposable
    {
        public enum PeerState
        {
            Invalid,
            Created,
            Signaling,
            SignalingFailed,
            Connected,
            Closing,
            Closed
        }

        private enum WebRtcInternalState
        {
            None,
            Signaling,
            SignalingFailed,
            Connected,
            Closed
        }



        private NativeWebRtcNetworkFactory mFactory;

        private object mLock = new object();

        private PeerState mState;

        private WebRtcInternalState mRtcInternalState;

        protected RTCPeerConnection mPeer;

        protected PeerConnectionInterface.RTCOfferAnswerOptions mOfferOptions;

        private Queue<string> mIncomingSignalingQueue = new Queue<string>();

        private Queue<string> mOutgoingSignalingQueue = new Queue<string>();

        private bool mDidSendRandomNumber;

        private int mRandomNumberSent;

        public PeerState State
        {
            get
            {
                return this.mState;
            }
        }

        public AWebRtcPeer(string[] urls, NativeWebRtcNetworkFactory factory)
        {
            this.mFactory = factory;
            this.SetupPeer(urls);
            this.OnSetup();
            this.mState = PeerState.Created;
        }

        protected abstract void OnSetup();

        protected abstract void OnStartSignaling();

        protected abstract void OnCleanup();

        private void SetupPeer(string[] urls)
        {
            PeerConnectionInterface.RTCConfiguration config = new PeerConnectionInterface.RTCConfiguration();
            PeerConnectionInterface.IceServer server = new PeerConnectionInterface.IceServer();
            for (int i = 0; i < urls.Length; i++)
            {
                server.urls.Add(urls[i]);
            }
            config.servers.Add(server);
            this.mPeer = this.mFactory.NativeFactory.CreatePeerConnection(config);
            this.mPeer.OnIceCandidate = new Action<IceCandidateInterface>(this.OnIceCandidate);
            this.mPeer.OnIceConnectionChange = new Action<PeerConnectionInterface.IceConnectionState>(this.OnIceConnectionChange);
            this.mPeer.OnRenegotiationNeeded = new Action(this.OnRenegotiationNeeded);
            this.mPeer.OnSignalingChange = new Action<PeerConnectionInterface.SignalingState>(this.OnSignalingChange);
        }

        public void Dispose()
        {
            if (this.mPeer != null)
            {
                this.Cleanup();
            }
        }

        private void Cleanup()
        {
            if (this.mState == PeerState.Closed || this.mState == PeerState.Closing)
            {
                return;
            }
            this.mState = PeerState.Closing;
            this.OnCleanup();
            if (this.mPeer != null)
            {
                this.mPeer.Dispose();
            }
            this.mPeer = null;
            this.mState = PeerState.Closed;
        }

        public void Update()
        {
            if (this.mState != PeerState.Closed && this.mState != PeerState.Closing && this.mState != PeerState.SignalingFailed)
            {
                this.UpdateState();
            }
            if (this.mState == PeerState.Signaling || this.mState == PeerState.Created)
            {
                this.HandleIncomingSignaling();
            }
        }

        private void UpdateState()
        {
            if (this.mRtcInternalState == WebRtcInternalState.Closed)
            {
                this.Cleanup();
                return;
            }
            if (this.mRtcInternalState == WebRtcInternalState.SignalingFailed)
            {
                this.mState = PeerState.SignalingFailed;
                return;
            }
            if (this.mRtcInternalState == WebRtcInternalState.Connected)
            {
                this.mState = PeerState.Connected;
            }
        }

        public void StartSignaling()
        {
            this.OnStartSignaling();
            this.CreateOffer();
        }

        public void NegotiateSignaling()
        {
            int nb = new System.Random(Guid.NewGuid().GetHashCode()).Next();
            this.mRandomNumberSent = nb;
            this.mDidSendRandomNumber = true;
            this.EnqueueOutgoing(string.Concat(nb));
        }

        public void AddSignalingMessage(string msg)
        {
            this.mIncomingSignalingQueue.Enqueue(msg);
        }

        public bool DequeueSignalingMessage(out string msg)
        {
            msg = null;
            object obj = this.mLock;
            bool result;
            lock (obj)
            {
                if (this.mOutgoingSignalingQueue.Count > 0)
                {
                    msg = this.mOutgoingSignalingQueue.Dequeue();
                    result = true;
                }
                else
                {
                    msg = null;
                    result = false;
                }
            }
            return result;
        }

        private void HandleIncomingSignaling()
        {
            while (this.mIncomingSignalingQueue.Count > 0)
            {
                string msg = this.mIncomingSignalingQueue.Dequeue();
                int randomNumber;
                if (int.TryParse(msg, out randomNumber))
                {
                    if (this.mDidSendRandomNumber)
                    {
                        if (randomNumber < this.mRandomNumberSent)
                        {
                            SLog.L("Signaling negotiation complete. Starting signaling.", new string[0]);
                            this.StartSignaling();
                        }
                        else if (randomNumber == this.mRandomNumberSent)
                        {
                            this.NegotiateSignaling();
                        }
                        else
                        {
                            SLog.L("Signaling negotiation complete. Waiting for signaling.", new string[0]);
                        }
                    }
                }
                else if (WebRtcWrap.IsIceCandidateJson(msg))
                {
                    IceCandidateInterface ice = WebRtcWrap.ParseJsonIceCandidate(msg);
                    if (ice != null)
                    {
                        this.mPeer.AddIceCandidate(ice);
                    }
                    else
                    {
                        SLog.LW("Invalid ice message received", new string[0]);
                    }
                }
                else if (WebRtcWrap.IsSessionDescriptionJson(msg))
                {
                    RTCSessionDescription sdi = RTCSessionDescription.FromJson(msg);
                    if (sdi != null)
                    {
                        if (sdi.type() == SessionDescriptionInterface.kOffer)
                        {
                            this.CreateAnswer(sdi);
                        }
                        else
                        {
                            this.RecAnswer(sdi);
                        }
                    }
                    else
                    {
                        SLog.LW("Invalid session description message received", new string[0]);
                    }
                }
                else
                {
                    SLog.LW("Invalid signaling message received", new string[0]);
                }
            }
        }

        private void EnqueueOutgoing(string msg)
        {
            object obj = this.mLock;
            lock (obj)
            {
                SLog.L("Signaling message " + msg, new string[0]);
                this.mOutgoingSignalingQueue.Enqueue(msg);
            }
        }

        private void CreateOffer()
        {
            this.mState = PeerState.Signaling;
            SLog.L("CreateOffer", new string[0]);
            this.mPeer.CreateOffer(delegate (RTCSessionDescription desc)
            {
                string msg = desc.ToJson();
                this.mPeer.SetLocalDescription(desc, delegate
                {
                    this.RtcSetSignalingStarted();
                    this.EnqueueOutgoing(msg);
                }, delegate (string error)
                {
                    SLog.LE(error, new string[0]);
                    this.RtcSetSignalingFailed();
                });
            }, delegate (string error)
            {
                SLog.LE(error, new string[0]);
                this.RtcSetSignalingFailed();
            }, this.mOfferOptions);
        }

        private void CreateAnswer(RTCSessionDescription offer)
        {
            this.mState = PeerState.Signaling;
            SLog.L("CreateAnswer", new string[0]);
            this.mPeer.SetRemoteDescription(offer, delegate
            {
                this.mPeer.CreateAnswer(delegate (RTCSessionDescription desc)
                {
                    string msg = desc.ToJson();
                    this.mPeer.SetLocalDescription(desc, delegate
                    {
                        this.RtcSetSignalingStarted();
                        this.EnqueueOutgoing(msg);
                    }, delegate (string error)
                    {
                        SLog.LE(error, new string[0]);
                        this.RtcSetSignalingFailed();
                    });
                }, delegate (string error)
                {
                    SLog.LE(error, new string[0]);
                    this.RtcSetSignalingFailed();
                }, null);
            }, delegate (string error)
            {
                SLog.LE(error, new string[0]);
                this.RtcSetSignalingFailed();
            });
        }

        private void RecAnswer(RTCSessionDescription answer)
        {
            SLog.L("RecAnswer", new string[0]);
            RTCPeerConnection rtcPeerConnection = this.mPeer;
            rtcPeerConnection.SetRemoteDescription(answer, new Action(delegate () { Debug.Log("Remote Description success"); }), delegate (string error)
             {
                 SLog.LE(error, new string[0]);
                 this.RtcSetSignalingFailed();
             });
        }

        private void RtcSetSignalingStarted()
        {
            if (this.mRtcInternalState == WebRtcInternalState.None)
            {
                this.mRtcInternalState = WebRtcInternalState.Signaling;
            }
        }

        protected void RtcSetSignalingFailed()
        {
            Debug.LogError("AWebRtcPeer RtcSignalingFailed");
            this.mRtcInternalState = WebRtcInternalState.SignalingFailed;
        }

        protected void RtcSetConnected()
        {
            if (this.mRtcInternalState == WebRtcInternalState.Signaling)
            {
                this.mRtcInternalState = WebRtcInternalState.Connected;
            }
        }

        protected void RtcSetClosed()
        {
            if (this.mRtcInternalState == WebRtcInternalState.Connected)
            {
                this.mRtcInternalState = WebRtcInternalState.Closed;
            }
        }

        private void OnIceCandidate(IceCandidateInterface candidate)
        {
            string msg = WebRtcWrap.ToJson(candidate);
            this.EnqueueOutgoing(msg);
        }

        private void OnIceConnectionChange(PeerConnectionInterface.IceConnectionState new_state)
        {
            SLog.L(new_state, new string[0]);
            if (new_state == PeerConnectionInterface.IceConnectionState.kIceConnectionFailed)
            {
                this.mState = PeerState.SignalingFailed;
            }
        }

        private void OnIceGatheringChange(PeerConnectionInterface.IceGatheringState new_state)
        {
            SLog.L(new_state, new string[0]);
        }

        private void OnRenegotiationNeeded()
        {
        }

        private void OnSignalingChange(PeerConnectionInterface.SignalingState new_state)
        {
            SLog.L(new_state, new string[0]);
            if (new_state == PeerConnectionInterface.SignalingState.kClosed)
            {
                this.RtcSetClosed();
            }
        }
    }
}
